qiskit_experiments/library/randomized_benchmarking/sampling_utils.py

```diff
@@ -0,0 +1,359 @@
+                selected_edges.append(
+                    rand_edge
+                )  # move random edge from all_edges to selected_edges
+                old_all_edges = all_edges[:]
+                all_edges = []
+                # only keep edges in all_edges that do not share a vertex with rand_edge
+                for edge in old_all_edges:
+                    if rand_edge[0] not in edge and rand_edge[1] not in edge:
+                        all_edges.append(edge)
+
+            two_qubit_prob = 0
+            try:
+                # need to divide by 2 since each two-qubit gate spans two lattice sites
+                two_qubit_prob = num_qubits * two_qubit_gate_density / 2 / len(selected_edges)
+            except ZeroDivisionError:
+                warnings.warn("Device has no connectivity. All gates will be single-qubit.")
+            if two_qubit_prob > 1 and not np.isclose(two_qubit_prob, 1):
+                warnings.warn(
+                    "Mean number of two-qubit gates is higher than the number of selected edges. "
+                    + "Actual density of two-qubit gates will likely be lower than input density."
+                )
+
+            put_1q_gates = set(qubits)
+            # put_1q_gates is a list of qubits that aren't assigned to a 2-qubit gate
+            # 1-qubit gates will be assigned to these edges
+            layer = []
+            for edge in selected_edges:
+                if self._rng.random() < two_qubit_prob:
+                    # with probability two_qubit_prob, place a two-qubit gate from the
+                    # gate set on edge in selected_edges
+                    if len(gateset[2][0]) == 1:
+                        layer.append(GateInstruction(tuple(edge), gateset[2][0][0]))
+                    else:
+                        layer.append(
+                            GateInstruction(
+                                tuple(edge),
+                                self._rng.choice(
+                                    np.array(gateset[2][0], dtype=Instruction),
+                                    p=[x / norm2q for x in gateset[2][1]],
+                                ),
+                            ),
+                        )
+                    # remove these qubits from put_1q_gates
+                    put_1q_gates.remove(edge[0])
+                    put_1q_gates.remove(edge[1])
+            for q in put_1q_gates:
+                if sum(gateset[1][1]) > 0:
+                    layer.append(
+                        GateInstruction(
+                            (q,),
+                            self._rng.choice(
+                                np.array(gateset[1][0], dtype=Instruction),
+                                p=[x / norm1q for x in gateset[1][1]],
+                            ),
+                        ),
+                    )
+                else:  # edge case of two qubit density of 1 where we still fill gaps
```

Comment:
Yes, the user should always provide a 1q gate distribution when using edge grab since it's probabilistic and is not guaranteed to maximally pack the layers with 2q gates. I've added error handing to check this.